### PR TITLE
ch4: replace all remaining POBJ locking with VCI locking

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

At this point, per VCI locking has fully replaced the POBJ locking. This PR swaps the last two remaining POBJ locks with VCI locking.

This PR fixes a bug revealed during PR #5033 testing:
```
not ok  - ./threads/rma/multifence 3
 
  Date: "Tue Feb  9 20:47:36 2021"
  ...
## Test output (expected 'No Errors'):
## multifence: src/mpid/ch4/src/ch4_init.c:990: MPID_Free_mem: Assertion `container' failed.
```

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

File: src/mpid/ch4/src/ch4_init.c [hunk 1]
```diff
@@ -526,29 +526,11 @@ int MPID_Init_local(int requested, int *provided)
...
-    MPIR_Assert(err == 0);
-#endif
-
-    MPID_Thread_mutex_create(&MPIDIU_THREAD_DYNPROC_MUTEX, &err);
-    MPIR_Assert(err == 0);
+    for (int i = 0; i < MAX_CH4_MUTEXES; i++) {
+        int err;
+        MPID_Thread_mutex_create(&MPIDI_global.m[i], &err);
+        MPIR_Assert(err == 0);
+    }
```
Comment (minsii) [inline src/mpid/ch4/src/ch4_init.c:531]: Compared with the original code, `MPIDIU_THREAD_HCOLL_MUTEX` is now always initialized no matter `HAVE_LIBHCOLL` is defined or not. But I think it is fine as it is just a mutex which is initialized but never used. To reduce potential concern from other reader, maybe we could have a brief comment here to explain such cases.

**Review by minsii — COMMENTED** [review]
The PR looks OK to me. I left two minor comments (one is a question for better understanding)

File: src/mpid/ch4/src/ch4_init.c [hunk 2]
```diff
@@ -985,11 +985,11 @@ int MPID_Free_mem(void *user_buf)
 
     alloc_mem_container_s *container;
 
-    MPID_THREAD_CS_ENTER(POBJ, MPIDIU_THREAD_ALLOC_MEM_MUTEX);
+    MPID_THREAD_CS_ENTER(VCI, MPIDIU_THREAD_ALLOC_MEM_MUTEX);
     HASH_FIND_PTR(alloc_mem_container_list, &user_buf, container);
     assert(container);
     HASH_DEL(alloc_mem_container_list, container);
-    MPID_THREAD_CS_EXIT(POBJ, MPIDIU_THREAD_ALLOC_MEM_MUTEX);
+    MPID_THREAD_CS_EXIT(VCI, MPIDIU_THREAD_ALLOC_MEM_MUTEX);
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/src/ch4_init.c:957]: Per-OBJ and per-VCI locking are essentially the same ideas -- fine-grained locking. The key difference is per-OBJ assigns a different lock to each object such as each communicator. It effectively addresses the parallelism inside MPICH, but when it goes down the network communication layer, it still needs to compete for finite network resources (e.g. channels). Per-object model lacks the mapping from each object to finite network interfaces. Per-VCI directly address this by assigning communication objects (mainly requests and communicators) explicitly to each VCI, thus solving the network mapping problems.

For non-communication critical components, per-object model and per-vci model are identical. However, historically both POBJ and per-VCI are using a mixed global/fine-grained locks in different parts, and in the beginning, the VCI implementation was trying to co-exist with per-OBJ and Global locking, thus creating rather a messy situation. The apparent ad-hoc conversion from POBJ to per-VCI was due to this code mess. At this point, we no longer have global locking in per-vci. All POBJ in ch4 should be replaced by per-VCI locks. POBJ is still used inside ch3 (no VCI). Above in MPIR-layer, you'll see both POBJ and per VCI locking in many places due to this ch3/ch4 situation.

File: src/mpid/ch4/src/ch4_init.c (hunk 1 above)
Comment (hzhou) [reply src/mpid/ch4/src/ch4_init.c:535]: Comment added.

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Review by minsii — APPROVED** [review]

**Pull request merged**